src/Storages/MergeTree/MergeTreeWhereOptimizer.cpp

```diff
@@ -193,8 +195,8 @@ void MergeTreeWhereOptimizer::analyzeImpl(Conditions & res, const ASTPtr & node,
             /// Condition depend on some column. Constant expressions are not moved.
             !cond.identifiers.empty()
             && !cannotBeMoved(node, is_final)
-            /// Do not take into consideration the conditions consisting only of the first primary key column
-            && !hasPrimaryKeyAtoms(node)
+            /// Do not take into consideration the conditions consisting of column that not belong to the primary key columns
+            && isColumnAllPrimaryKey(node)
```

Comment:
Hm, this change is unclear to me.
Previously, we did not allow to move condition to PREWHERE in case this condition was an expression over *first* PK column. Like, if we had PK `(a, b, c)` then we could not move condition `where a = 42`, but could move `where b = 42` and `where another_column = 42`. This logic is understandable : most of the data would be filtered by PK if condition contains the first PK column (and this is mostly not true for other PK columns). Actually, I thought that maybe we can remove this check - because even for the condition over first PK column we will likely read 1-2 granules per part.

Now, we allow to move condition *only* if it is expression over PK key. Maybe we should check for FINAL as well? Something like ` && (!is_final || isColumnAllPrimaryKey(node))`

Reply:
btw, why here we use PK, not sorting key? Final requires sorting key to be processed, not just PK (which is a prefix of sorting key). PK is a set of columns for which we write an index.

Reply:
>if we had `PK (a, b, c) `then we could not move condition `where a = 42`, but could move `where b = 42` and `where another_column = 42`. . This logic is understandable : most of the data would be filtered by PK if condition contains the first PK column

My situation: A ReplacingMergeTree table `event_share_data` has 39 columns, and it is  ORDER BY `mongo_id` column. And it has 10819183 records.
My query:
```
select mongo_id ,  subtype ,    image_path ,   meta_type,  type,   subtype,    title
    FROM event_share_data  FINAL
    WHERE mongo_id IN ('95f344d7c7033989bba9880414026463', '9778a571fb553748927e895a7df93581', '9c645702c7933102b4aa318fb8cc6b2d', '51d25df5b6083a0989dbaee3bac09329', '54c60154557b3f098c2b8f2043b21096', '554d46fb767e3917b01f3c67d26fbbd4', '585a154adbf03832b22dd381cc0f6f44', '5aba1f84c62134ba82161216323cbdf1', '5de5d9875ca13f5c8d781d53e0a438c2', '3b4093d42d3f3516a299c265b3b06688', '3edd4fbce96435c6addd56cb5d57fcba', '40b111c3217b3fa88fbf618ade389e26', '4189c51eeb84333abf09a3519469ece5', '425c8088630737d29d717b248687d3fc', '430ef1206ce03672be50f40a06a13c55', '434b9d3e4300333aa12bf6aa9a3dffd5', '0767201926f730fa971fa4fa2dc5f242', '6848ecf95a8939758e54f4ea52842d4e', '68dde059f19435d0aa446e0019ef4f22', '6ce19440486b3386871e15a6cbde2712', '6d0667b1846531519fac98b9e53dbeef', '6dbb4d9545703ce8b5609c5e5edd1123', '6e645b9c4b62333583b79346865d0ad6', '6e9ce3f825183cc9a0f768cf7475d7e4', '73e184ce84d03930a4a7631987774cb4', '7f6277c322473ac3be9db547604261f2', '817a1d3a44dc333c9ada1df6179c5700', '8641d2207df2374eabb3243ddd4b3c61', '8777a28ec7cf32219a142125869a6af4', '87effcb49ceb3a3a9df1106888c1c06b')
```


I found that in my situation, if `where mongo_id in (...)` can be optimize to `prewhere`, then the execute time of this query can be reduced from 2.075sec. to 0.384sec.
So in my opnion, even the first primary key/sorting key should be optimize to `prewhere`,  it can improve the performace.

Reply:
fixed.